File: ndb/questions/01_greedy/question06/main2.java
import java.util.*;

public class Main {

    static class Solution {

        public int solution(int[] food_times, long k) {
            List<Food> foods = new LinkedList<>();
            // 음식의 총 개수
            int total = food_times.length;
            // LinkedList 에 추가
            for (int i = 0; i < total; i++){
                // (음식 시간, 음식 번호)
                foods.add(new Food(food_times[i], i + 1));
            }
            // 음식 시간기준 오름차순 정렬
            foods.sort(Comparator.comparingInt(o -> o.time));
            // 이전 시간
            int previousTime = 0;
            // 현재 인덱스
            int currentIndex = 0;
            for (Food f : foods) {
                // column 의 높이
                long diff = f.time - previousTime;
                System.out.println("diff = " + diff);
                if (diff != 0) {
                    // column.length * row.length
                    long spend = diff * total;
                    System.out.println("spend = " + spend);
                    if (spend <= k) {
                        k -= spend;
                        previousTime = f.time;
                        System.out.println("k = " + k);
                        System.out.println("previousTime = " + previousTime);
                    } else {
                        // 남은 음식 index 기준 재정렬
                        k %= total;
                        foods.subList(currentIndex, foods.size()).sort(Comparator.comparingInt(o -> o.index));
                        System.out.println("---result---");
                        System.out.println("k = " + k);
                        System.out.println("currentIndex = " + currentIndex);
                        // 다음 먹어야할 음식 연산
                        return foods.get(currentIndex + (int) k).index;
                    }
                }
                currentIndex++;
                total--;
            }

            return -1;
        }

        static class Food {
            int time;
            int index;

            public Food(int time, int index) {
                this.time = time;
                this.index = index;
            }
        }
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        int[] array = {3, 1, 2};
        System.out.println(solution.solution(array, 5));
    }
}
/*
diff = 1
spend = 3
k = 2
previousTime = 1
diff = 1
spend = 2
k = 0
previousTime = 2
diff = 1
spend = 1
---result---
k = 0
currentIndex = 2
1
*/

File: ndb/questions/01_greedy/question06/main2.dart
class Solution {
  int solution(List<int> foodTimes, double k) {
    List<Food> foods = <Food>[];
    // 음식의 총 개수
    int total = foodTimes.length;
    // List 추가
    for (int i = 0; i < total; i++) {
      // (음식 시간, 음식 번호)
      foods.add(Food(time: foodTimes[i], index: i + 1));
    }
    // 음식 시간기준 오름차순 정렬
    foods.sort((a, b) => a.time.compareTo(b.time));
    // 이전 시간
    int previousTime = 0;
    // 현재 인덱스
    int currentIndex = 0;
    for (Food food in foods) {
      // column 높이
      int diff = food.time - previousTime;
      print('diff = $diff');
      if (diff != 0) {
        // column.length * row.length
        int spend = diff * total;
        print('spend = $spend');
        if (spend <= k) {
          k -= spend;
          previousTime = food.time;
          print('k = $k');
          print('previousTime = $previousTime');
        } else {
          k %= total;
          foods
              .sublist(currentIndex, foods.length)
              .sort((a, b) => a.index.compareTo(b.index));
          print('---result---');
          print('k = $k');
          print('currentIndex = $currentIndex');
          // 다음 먹어야할 음식 연산
          return foods[currentIndex + k.toInt()].index;
        }
      }
      currentIndex++;
      total--;
    }
    return -1;
  }
}

class Food {
  final int time;
  final int index;

  Food({required this.time, required this.index});
}

void main(List<String> arguments) {
  Solution solution = Solution();
  print(solution.solution([3, 1, 2], 5));
}
/*
diff = 1
spend = 3
k = 2.0
previousTime = 1
diff = 1
spend = 2
k = 0.0
previousTime = 2
diff = 1
spend = 1
---result---
k = 0.0
currentIndex = 2
1
*/